### DISCUSSION TITLE
How to strip milliseconds from Date serialization in response

### DISCUSSION THREAD

**Discussion by clementluxair** [fastify/fastify#5687] [Q&A] [answered]
As stated in [documentation](https://fastify.dev/docs/v4.0.x/Reference/Reply/#senddata), Fastify uses fast-json-stringify to serialize JavaScript objects before they are sent back to client. During that process, Date instances are serialized with milliseconds (ex: 2024-09-17T07:42:01.296Z).

Is there a way to configure Fastify or fast-json-stringify to strip them, without having to write a custom serializer (which would make me lose the performance benefit of fast-json-stringify)?

**Comment by chrisjcthomas** [comment, marked as answer]
There is no Fastify / `fast-json-stringify` knob that prints `Date` as `YYYY-MM-DDTHH:mm:ssZ` (no fractional seconds). With a response schema, FJS treats `Date` via `toISOString()`, so you always get milliseconds for `format: 'date-time'`.

You can still keep FJS (and your schemas) — just hand it **strings** instead of `Date` instances, or rewrite dates in `preSerialization` before FJS runs.

### 1. Format in the handler (simplest, still uses FJS)

Schema field stays `type: 'string'` / `format: 'date-time'`. Convert before `return`:

```js
const toDateTimeNoMs = (d) =>
  (d instanceof Date ? d : new Date(d)).toISOString().replace(/\.\d{3}Z$/, "Z")

fastify.get("/item", {
  schema: {
    response: {
      200: {
        type: "object",
        properties: {
          createdAt: { type: "string", format: "date-time" },
        },
      },
    },
  },
}, async () => ({
  createdAt: toDateTimeNoMs(new Date()),
}))
```

FJS will not call `toISOString()` again on a plain string; it just escapes/emits it. You keep the compiled serializer.

### 2. App-wide `preSerialization` hook

Useful when many handlers still return `Date` objects:

```js
function stripDateMs(value) {
  if (value instanceof Date) {
    return value.toISOString().replace(/\.\d{3}Z$/, "Z")
  }
  if (Array.isArray(value)) return value.map(stripDateMs)
  if (value && typeof value === "object") {
    const out = {}
    for (const [k, v] of Object.entries(value)) out[k] = stripDateMs(v)
    return out
  }
  return value
}

fastify.addHook("preSerialization", async (_req, _reply, payload) => {
  return stripDateMs(payload)
})
```

Hook runs **before** the response schema serializer, so FJS still does the fast stringify on the transformed payload.

### 3. What will not work

- Expecting `format: 'date-time'` to drop ms — it does not
- `format: 'date'` / `'time'` — different shapes (day-only / time-only)
- Dropping the response schema so you can `JSON.stringify` yourself — you lose FJS and schema-driven serialization

There is no global “date format = seconds precision” option in Fastify today. Pre-format strings (handler or `preSerialization`) is the usual approach and does not force you off `fast-json-stringify`.

**Reply by clementluxair** [reply]
Thank you for the suggestions.
